### PR TITLE
OpticalFlow: Add support for MinnowBoardMax [v2]

### PR DESCRIPTION
**Pull request by ricardotk**
This PR adds support for OpticalFlow to MinnowBoardMax trying to
leave the OpticalFlow implementation as generic as it already is.

We had to add some format conversion and software crop to the cameras that
do not have this features.

Most cameras do not support NV12 or GREY formats, we are adding in this PR
a conversion from YUYV format, that seems pretty common in cameras, to GREY
format (since we do not use Cb and Cr data on OpticalFlow).

Also we are adding a software crop for 8bpp images such as GREY format. For the
same reason, most cameras do not have support for overlaying (crop, resize and so on).

This PR is still a WIP version since it still need to be tested on a flying drone, but the tests I had made with a camera seemed to be correct.

* Differences from v1:
  * Changed `crop_image_8bpp()` algorithm
  * Correct check to see if crop is supported
  * Remove unused `get_fd()` function
  * Adopt 80~100 columns coding style

### CODE REVIEW

**Comment by jberaud** [comment]
I am ok with your modifications. I will probably add a condition for checking the return of the S_CROP ioctl in order for you not to have a message printed at every startup.

For the rest, I'll leave it to @lucasdemarchi especially regarding the use of alloca for video buffers, since if someone had the idea to increase sizes, they could easily be too big for the stack and then alloca has an undefined behaviour. If Lucas is ok with that, I'll add the reviewed flag.

**Comment by ricardotk** [comment]
Sounds good, thanks a lot for the review Julien (:

File: libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp [hunk 1]
```diff
@@ -71,18 +84,10 @@ void OpticalFlow_Onboard::init(AP_HAL::OpticalFlow::Gyro_Cb get_gyro)
...
-    crop_width = HAL_OPTFLOW_ONBOARD_CROP_WIDTH;
-    crop_height = HAL_OPTFLOW_ONBOARD_CROP_HEIGHT;
-    top = 0;
-    /* make the image square by cropping to YxY, removing the lateral edges */
-    left = (HAL_OPTFLOW_ONBOARD_SENSOR_WIDTH -
-            HAL_OPTFLOW_ONBOARD_SENSOR_HEIGHT) / 2;
     _format = V4L2_PIX_FMT_NV12;
+#elif CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_MINLURE
+    _format = V4L2_PIX_FMT_YUYV;
+#endif
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp]: In order for this to be more generic, couldn't we query the supported formats from the camera we are using and set to one of the formats that we support?  You picked a random usb camera, other people may have other cameras with different formats.

Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp]: A quick search: http://linuxtv.org/downloads/v4l-dvb-apis/vidioc-enum-fmt.html

Comment (jberaud) [inline libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp]: Good idea.

**Comment by lucasdemarchi** [comment]
> For the rest, I'll leave it to @lucasdemarchi especially regarding the use of alloca for video buffers, since if someone had the idea to increase sizes, they could easily be too big for the stack and then alloca has an undefined behaviour. If Lucas is ok with that, I'll add the reviewed flag.

I prefer to maintain a pointer in the class and allocate the buffer during initialization. This also means we don't need some checks for `crop_by_software`

**Comment by ricardotk** [comment]
> I prefer to maintain a pointer in the class and allocate the buffer during initialization. This also means we don't need some checks for crop_by_software.

This way I would have to store the sizes as the class attributes as well, wouldn't it be better to `malloc()` this buffers in the beginning of `_run_optflow()` and `free()` them after they exit the loop?

**Comment by ricardotk** [comment]
#3459 v3

**Comment by lucasdemarchi** [comment]
Ok
Em 8 de jan de 2016 11:01 AM, "Ricardo de Almeida Gonzaga" <
notifications@github.com> escreveu:

**Pull request closed without merging**